ESP8266Interface.cpp

```diff
@@ -472,6 +478,33 @@ void ESP8266Interface::socket_attach(void *handle, void (*callback)(void *), voi
     _cbs[socket->id].data = data;
 }
 
+nsapi_error_t ESP8266Interface::setsockopt(nsapi_socket_t handle, int level,
+        int optname, const void *optval, unsigned optlen)
+{
+    struct esp8266_socket *socket = (struct esp8266_socket *)handle;
+
+    if(socket->connected == true) {// ESP8266 limitation, keepalive needs to be given before connecting
+        return NSAPI_ERROR_UNSUPPORTED;
+    }
+
+    if (level == NSAPI_SOCKET) {
+        switch (optname) {
+            case NSAPI_KEEPALIVE: {
```

Comment:
Probably should say "unsupported" if it's not a TCP socket, I guess.

Reply:
ok

Reply:
Yep, the NSAPI doesn't use its level enum properly - all options are that one level, rather than being Socket/IP/TCP levels.

But there's a passthrough mechanism - if you know the underlying stack, you can pass its level/option numbers to access its native socket options.

But my point was that even without a proper level/proto distinguisher you can be checking `socket->proto` in the particular TCP-specific optname case(s).

Reply:
ok, I'll add the check